drivers/sensor/scd30/scd30.c

```diff
@@ -0,0 +1,375 @@
+
+	rc = scd30_write_command(dev, SCD30_CMD_READ_MEASUREMENT);
+	if (rc != 0) {
+		LOG_DBG("Failed to send command. (rc = %d)", rc);
+		return rc;
+	}
+
+	/* delay for 3 msec as per datasheet. */
+	k_sleep(K_MSEC(3));
+
+	rc = i2c_read_dt(&cfg->bus, (uint8_t *)&raw_rx_data, sizeof(raw_rx_data));
+	if (rc != 0) {
+		LOG_DBG("Failed to read data. (rc = %d)", rc);
+		return rc;
+	}
+
+	/* C02 data */
+	rc = scd30_fill_data_buf(raw_rx_data.co2_msw, &rx_data.co2_be[0]);
+	if (rc != 0) {
+		return rc;
+	}
+	rc = scd30_fill_data_buf(raw_rx_data.co2_lsw, &rx_data.co2_be[SCD30_WORD_SIZE]);
+	if (rc != 0) {
+		return rc;
+	}
+
+	/* temperature data */
+	rc = scd30_fill_data_buf(raw_rx_data.temp_msw, &rx_data.temp_be[0]);
+	if (rc != 0) {
+		return rc;
+	}
+	rc = scd30_fill_data_buf(raw_rx_data.temp_lsw, &rx_data.temp_be[SCD30_WORD_SIZE]);
+	if (rc != 0) {
+		return rc;
+	}
+
+	/* relative humidity */
+	rc = scd30_fill_data_buf(raw_rx_data.humidity_msw, &rx_data.humidity_be[0]);
+	if (rc != 0) {
+		return rc;
+	}
+	rc = scd30_fill_data_buf(raw_rx_data.humidity_lsw, &rx_data.humidity_be[SCD30_WORD_SIZE]);
+	if (rc != 0) {
+		return rc;
+	}
+
+	data->co2_ppm = scd30_bytes_to_float(rx_data.co2_be);
+	data->temp = scd30_bytes_to_float(rx_data.temp_be);
+	data->rel_hum = scd30_bytes_to_float(rx_data.humidity_be);
+
+	return 0;
+}
+
+static int scd30_channel_get(const struct device *dev, enum sensor_channel chan,
+			     struct sensor_value *val)
+{
+	struct scd30_data *data = dev->data;
+
+	switch (chan) {
+	case SENSOR_CHAN_CO2:
+		sensor_value_from_double(val, data->co2_ppm);
+		break;
+
+	case SENSOR_CHAN_AMBIENT_TEMP:
+		sensor_value_from_double(val, data->temp);
+		break;
+
+	case SENSOR_CHAN_HUMIDITY:
+		sensor_value_from_double(val, data->rel_hum);
+		break;
+
+	default:
+		return -1;
+	}
+	return 0;
+}
+
+static int scd30_attr_get(const struct device *dev, enum sensor_channel chan,
+			  enum sensor_attribute attr, struct sensor_value *val)
+{
+	struct scd30_data *data = dev->data;
+
+	if (chan != SENSOR_CHAN_ALL) {
+		return -ENOTSUP;
+	}
+
+	switch (attr) {
+	case SENSOR_ATTR_SAMPLING_FREQUENCY:
+	{
+		float frequency = 1.0 / data->sample_time;
+
+		sensor_value_from_double(val, frequency);
+		return 0;
+	}
+
+	default:
+		return -ENOTSUP;
+	}
+}
+
+static int scd30_attr_set(const struct device *dev, enum sensor_channel chan,
+			  enum sensor_attribute attr, const struct sensor_value *val)
+{
+	if (chan != SENSOR_CHAN_ALL) {
+		return -ENOTSUP;
+	}
+
+	switch (attr) {
+	case SENSOR_ATTR_SAMPLING_FREQUENCY:
+	{
+		double frequency = sensor_value_to_double(val);
+		double sample_time = 1.0 /  frequency;
+
+		return scd30_set_sample_time(dev, (uint16_t)sample_time);
+	}
+	default:
+		return -ENOTSUP;
+	}
+}
+
+static const struct sensor_driver_api scd30_driver_api = {
+	.sample_fetch = scd30_sample_fetch,
+	.channel_get = scd30_channel_get,
+	.attr_get = scd30_attr_get,
+	.attr_set = scd30_attr_set,
+};
+
+static int scd30_init(const struct device *dev)
```

Comment:
suggestion: initial sampling period could be set in Devicetree and initialized here.